File: benchmark/languages/go/track_a/k_nucleotide.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func normalizeSequence(path string) (string, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return "", err
	}
	var builder strings.Builder
	for _, line := range strings.Split(string(data), "\n") {
		trimmed := strings.TrimSpace(line)
		if trimmed != "" && !strings.HasPrefix(trimmed, ">") {
			builder.WriteString(strings.ToUpper(trimmed))
		}
	}
	return builder.String(), nil
}

func countKmers(sequence string, k int) map[string]int {
	counts := map[string]int{}
	if len(sequence) < k {
		return counts
	}
	for i := 0; i <= len(sequence)-k; i++ {
		token := sequence[i : i+k]
		counts[token]++
	}
	return counts
}

func main() {
	path := "benchmark/data/track_a/dna_input.fasta"
	if len(os.Args) > 1 {
		path = os.Args[1]
	}
	sequence, err := normalizeSequence(path)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}

	freq1 := countKmers(sequence, 1)
	freq2 := countKmers(sequence, 2)
	freq3 := countKmers(sequence, 3)
	freq4 := countKmers(sequence, 4)
	freq6 := countKmers(sequence, 6)
	freq8 := countKmers(sequence, 8)

	fmt.Println(len(sequence))
	fmt.Println(len(freq1))
	fmt.Println(len(freq2))
	fmt.Println(len(freq3))
	fmt.Println(len(freq4))
	fmt.Println(len(freq6))
	fmt.Println(len(freq8))
	fmt.Println(freq3["GGT"])
	fmt.Println(freq4["GGTA"])
	fmt.Println(freq6["GGTATT"])
	fmt.Println(freq8["AGGGTAAA"])
}

File: benchmark/languages/rust/track_a/k_nucleotide.rs
use std::collections::HashMap;
use std::env;
use std::fs;

fn normalize_sequence(path: &str) -> Result<String, std::io::Error> {
    let raw = fs::read_to_string(path)?;
    let mut out = String::with_capacity(raw.len());
    for line in raw.lines() {
        let trimmed = line.trim();
        if !trimmed.is_empty() && !trimmed.starts_with('>') {
            out.push_str(&trimmed.to_ascii_uppercase());
        }
    }
    Ok(out)
}

fn count_kmers(sequence: &str, k: usize) -> HashMap<String, usize> {
    let mut counts = HashMap::new();
    if sequence.len() < k {
        return counts;
    }
    for i in 0..=(sequence.len() - k) {
        let token = &sequence[i..(i + k)];
        *counts.entry(token.to_string()).or_insert(0) += 1;
    }
    counts
}

fn main() {
    let path = env::args()
        .nth(1)
        .unwrap_or_else(|| "benchmark/data/track_a/dna_input.fasta".to_string());
    let sequence = match normalize_sequence(&path) {
        Ok(value) => value,
        Err(err) => {
            eprintln!("{}", err);
            std::process::exit(1);
        }
    };

    let freq1 = count_kmers(&sequence, 1);
    let freq2 = count_kmers(&sequence, 2);
    let freq3 = count_kmers(&sequence, 3);
    let freq4 = count_kmers(&sequence, 4);
    let freq6 = count_kmers(&sequence, 6);
    let freq8 = count_kmers(&sequence, 8);

    println!("{}", sequence.len());
    println!("{}", freq1.len());
    println!("{}", freq2.len());
    println!("{}", freq3.len());
    println!("{}", freq4.len());
    println!("{}", freq6.len());
    println!("{}", freq8.len());
    println!("{}", freq3.get("GGT").copied().unwrap_or(0));
    println!("{}", freq4.get("GGTA").copied().unwrap_or(0));
    println!("{}", freq6.get("GGTATT").copied().unwrap_or(0));
    println!("{}", freq8.get("AGGGTAAA").copied().unwrap_or(0));
}
